Implementation of Random Forest Nearest Neighbors (RF-NN)

grovduck opened the pull request:
As detailed in #24, the last major estimator that we wanted to port from `yaImpute` was Random Forest Nearest Neighbors (RF-NN) as originally developed by Crookston and Finley.  This PR introduces a scikit-learn compatible estimator for using this approach.  

While we have tried to stay as consistent as possible with the original `R` implementation, some differences will be present in this implementation.  Rather than use R's `randomForest` library (through use of `rpy2`), we opt instead to use scikit-learn's [`RandomForestClassifier`](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html) and [`RandomForestRegressor`](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html) so that users can use a pure Python solution.  Differences in implementation exist between the R and Python objects and we will attempt to reproduce consistent (though not exact) results.

This PR will be multi-staged to verify correctness with the `yaImpute` implementation and will serve as a record of the steps we've conducted while transitioning to the `sklearn` estimators.

1. We will push a temporary commit that uses R and rpy2 and tests outputs from `yaImpute`.  We've captured the outputs for neighbors, distances, and predicted `y` values (both weighted and unweighted) from `yaImpute` and use these in the `test_port.py` suite of tests.  @grovduck and @aazuspan will run these tests locally to ensure they pass.  Note that these tests won't pass on CI as we don't have R setup and we don't want to go to the effort of setting this up for a temporary solution.
2. Once we've verified that these local tests are passing, remove the rpy2 implementation and build in the `RandomForestClassifier` / `RandomForestRegressor` solution.  At this point, we will remove the neighbor and distance files from `yaImpute`, but retain the predicted values for each sample.  We will write a (temporary) test in `test_port.py` that verifies that the correlation coefficient for each `y` attribute from the R and Python outputs exceed a certain threshold.  These tests should pass CI as well as the dependency on `rpy2` will go away.  **EDIT**: There is more detail in [this comment](https://github.com/lemma-osu/sknnr/pull/85#issuecomment-2613658252), but ultimately we couldn't get the correspondence between R's `randomForest` and scikit-learn's `RandomForestRegressor` to match well enough to effectively test this.  This is outside our control and we decided to push forward with development.
3. Capture the neighbors/distances/predicted values using the sklearn estimators and replace those in the data files for `test_port.py`. Whereas all other `sknnr` estimators are from the original `yaImpute` ouput, these will be output from `sknnr` and will persist for the duration of this PR.  Eventually, all outputs ensuring port compatibility between `yaImpute` and `sknnr` will be replaced as proposed in #42.

Checklist:
- [x] Commit `rpy2` implementation which calls R directly and should match `yaImpute` output
- [x] Remove neighbor and distance files from `rpy2` implementation, but retain predicted files
- [x] Implement `sklearn` solution
- [x] Write test in `test_port.py` to ensure high correlations between `yaImpute` and `sknnr` output
- [x] Replace neighbor, distance, and prediction test files with output from `sknnr`
- [x] Iterate on `RFNNRegressor` design while ensuring "port" tests pass
- [x] Ensure `RFNNRegressor` passes all tests and uses `sklearn` tags correctly

grovduck commented:
@aazuspan, hopefully my convoluted plan here makes sense.  Unfortunately, I'm hoping to lean on you a bit for testing during this process and I hope it doesn't take too much setup for you.  

The implementation right now is using random forest regression trees for each `y` attribute from the Moscow Mountain / St. Joes dataset.  It's a horrible fit and not used in the examples given in `yaImpute`, but it does complete and gives the desired output.  You will receive a number of R warnings (through `rpy2`).  Currently `RFNNRegressor` is only being tested in `test_port.py` - I want to wait until steps 1 and 2 are finished (and have the `sklearn` implementation) before we start testing with the other packages.

Right now, the code is also not terribly well organized and there is a lot of repeated code that should go away in future iterations.  But here are some of the things I'm currently thinking about:

- I've been going back and forth as to whether it intuitively makes sense to have a transformer within the estimator (like we do for all other estimators beside `RawKNNRegressor`).  The "transform" here would be from the original `X` matrix (samples by features) to the node ID matrix (samples by forest/tree identifiers).  For example, if there were 10 forests and 100 trees per forest, there would be 1000 features in the output transform which capture the node IDs for each sample.  This sets it up nicely to mimic the data structures that we have for the other estimators, but I'm wondering if the idea of a transform is too esoteric since the node IDs are meaningless without the context of the actual forests themselves (although the same could be argued for the ordination methods).  I think we'd be able to take advantage of some of our existing base classes if we went this route.
- I'm really hoping we can use `kneighbors` from `TransformedKNeighborsRegressor` and use the `metric` keyword to implement the required distance finding for RFNN.  I haven't done enough investigation yet, but `metric` can take a [`DistanceMetric`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.DistanceMetric.html) object and compute pairwise distances in this way.  I'm looking at the Hamming distance and that might work?  If not, one can create a custom metric as long as it satisfies the four conditions set out on that page (although the triangle inequality one is throwing me off!)
- I'm guessing we'll want to allow `kwargs` to flow through to `RandomForestClassifier` and `RandomForestRegressor` in terms of specifying how the forests get built, but we'll also want to come up with reasonable defaults for trees per forests as this could explode fairly easily.
Let me know if anything else strikes you right away.  I'm feeling like this might take some time to get exactly right and I appreciate your patience already!

aazuspan commented:
@grovduck, wow! I'm very impressed by the detailed roadmap - looks like you've thought of everything here. It's exciting to see a working version of RF-NN, even if the RF part is a placeholder for now. I imagine this is the hard part done, at least.

The port tests ran perfectly on my machine (aside from the R warning about fit quality that you mentioned). I need to do a closer read of everything and take some time to give thoughtful responses to the implementation points you brought up, but I didn't want to block you since it sounds like we're ready to move onto step 2 with the `sklearn` RF implementation?

grovduck commented:
Awesome, thanks @aazuspan!  Glad this worked for you as well.  (As a side note, rather than squashing all commits on this PR, I wonder if it might be worth holding onto one with the `rpy2` implementation so that we can come back to it at some point?)

Looks like Hamming distance is a winner, so I'm going to try to get that implemented while I am still using `rpy2` and then start working on the `sklearn` estimators.
```
>>> import numpy as np
>>> from sklearn.metrics import DistanceMetric
>>> hamming = DistanceMetric.get_metric("hamming")

>>> # Set references and targets (like node IDs)
>>> refs = np.arange(24).reshape(6, 4)
>>> refs
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])
>>> targets = np.array([[0, 13, 14, 19], [4, 21, 18, 3]]) 
>>> targets
array([[ 0, 13, 14, 19],
       [ 4, 21, 18,  3]])

>>> # Find pairwise differences between targets and refs
>>> # 0 is perfect match, 1 is no matches
>>> # Returned as samples x refs array
>>> hamming.pairwise(targets, refs)
array([[0.75, 1.  , 1.  , 0.5 , 0.75, 1.  ],
       [0.75, 0.75, 1.  , 1.  , 0.75, 0.75]])

grovduck commented:
@aazuspan, I decided to push forward with the idea of a transformer to store the node IDs as it began to make a more conceptual sense to me and vastly reduced the code.  This most recent commit does the following:

- Isolate `rpy2` functions into separate module for easy removal (`src/sknnr/transformers/_rf_rpy2.py`)
- Create new `RFNodeTransformer` which inherits from `TransformerMixin` and is responsible for `fit` and `transform`.  I'm not at all sold on the name, but wanted to convey that it transforms from X to node IDs given previously fit random forests.
- Simplified `RFNNRegressor` by inheriting from `TransformedKNeighborsRegressor` and setting `metric="hamming"`.  This class now only has an `__init__` and a `_get_transformer`.

In redoing this way, our port tests for neighbors (and, subsequently, predict) are now failing.  As noted [on this page](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html),

> Regarding the Nearest Neighbors algorithms, if it is found that two neighbors, neighbor k+1 and k, have identical distances but different labels, the results will depend on the ordering of the training data.

What I've observed is, that in the case of ties, the larger ID is being recorded as a nearer neighbor in this code, whereas it's reversed in `yaImpute`'s case.  On the previous commit, I went through some gyrations to guarantee the lower ID is returned first to match `yaImpute`, but that seems arbitrary and not worth it for data we will soon toss.  By commenting out these lines:
```
assert_array_equal(nn, dataset.ref_neighbors)
assert_array_equal(nn, dataset.trg_neighbors)
```
and running `hatch run test:all tests/test_port.py::test_kneighbors -k randomForest --verbose`, hopefully tests will still pass for you.

aazuspan commented:
> Create new RFNodeTransformer which inherits from TransformerMixin and is responsible for fit and transform. I'm not at all sold on the name, but wanted to convey that it transforms from X to node IDs given previously fit random forests.

I like the name. `RFTransformer` would match the pattern of the other transformers that just use the estimator name, but if the "node" aspect is important then I think `RFNodeTransformer` is a good choice, especially if it might be used independent of the estimator.

> Simplified RFNNRegressor by inheriting from TransformedKNeighborsRegressor and setting metric="hamming". This class now only has an __init__ and a _get_transformer.

This is really clean! I love that it matches the design of the rest of our transformed estimators by abstracting out the RF terminal node matching as a transformation.

>  What I've observed is, that in the case of ties, the larger ID is being recorded as a nearer neighbor in this code, whereas it's reversed in yaImpute's case. On the previous commit, I went through some gyrations to guarantee the lower ID is returned first to match yaImpute, but that seems arbitrary and not worth it for data we will soon toss.

I'm not sure I fully understand why this changed with the introduction of Hamming distance or why we have tied neighbors, but you seem to have a good handle on it so I'm satisfied! I agree that if this is just an arbitrary difference in the tie-breaking strategy between `yaImpute` and `sklearn`, it's not worth the effort of making a workaround for a temporary test.

> and running hatch run test:all tests/test_port.py::test_kneighbors -k randomForest --verbose, hopefully tests will still pass for you.

Confirmed, still passing with the neighbor array checks commented out.

> As a side note, rather than squashing all commits on this PR, I wonder if it might be worth holding onto one with the rpy2 implementation so that we can come back to it at some point?

Yeah, rebasing this one makes sense to me.

It seems like this is coming along really well! The overall design looks great and I don't think I have any substantive comments at this point. I don't want to distract you with nits if things are still in progress, but let me know if there's anything that would be helpful at this point.

grovduck commented:
Thanks for the great feedback at this point, @aazuspan.  Exactly at the right level for now - a closer look will be really helpful as we close in on final-ish code, but verification and helping out with big-level design right now is very much appreciated.

> I like the name. `RFTransformer` would match the pattern of the other transformers that just use the estimator name, but if the "node" aspect is important then I think `RFNodeTransformer` is a good choice, especially if it might be used independent of the estimator.

I churned on the transformer name a bit as well.  `RFTransformer` felt like it might be giving the leaf node *values* rather than IDs, but I guess if we wanted to be very explicit, we could go with `RFNodeIDTransformer`.  I'll keep the name as-is for now and we can cycle back if warranted.

> This is really clean! I love that it matches the design of the rest of our transformed estimators by abstracting out the RF terminal node matching as a transformation.  

Yeah, I was really excited how clean it was as well.  That speaks to your careful architecture at the beginning!  Thanks for this.  The fact that we can mostly use `RawNNRegressor` (and really `KNearestRegressor`) for `kneighbors` and `predict` without too much effort is huge!

> I'm not sure I fully understand why this changed with the introduction of Hamming distance or why we have tied neighbors, but you seem to have a good handle on it so I'm satisfied!

I had ties in the previous iteration as well, but was using `np.argsort` to find the most matches and actually had to turn it around to find the most differences.  We could enforce the lower ID in the case of ties as a post-step to finding neighbors/distances, but most of that is being done in `KNearestRegressor.kneighbors` and we've already subsetted to the number of wanted neighbors by the time we get the neighbors/distances back.  

I'm actually a bit surprised that there were as many ties as there were, but Hamming is more of a discrete distance metric than n-dimensional Euclidean space.

> > As a side note, rather than squashing all commits on this PR, I wonder if it might be worth holding onto one with the rpy2 implementation so that we can come back to it at some point?
> 
> Yeah, rebasing this one makes sense to me.

I might need some help with this one when it comes time to do so.  Are you thinking an interactive rebase and picking one of the `rpy2` related commits and the final one?  We can cross that bridge when we get there.

aazuspan commented:
> I had ties in the previous iteration as well, but was using np.argsort to find the most matches and actually had to turn it around to find the most differences. 

Gotcha, thanks for the explanation!

> I'm actually a bit surprised that there were as many ties as there were, but Hamming is more of a discrete distance metric than n-dimensional Euclidean space.

Ah, I think I'm starting to understand now. So Hamming distance is just calculating the percent of node IDs that are mismatched between samples? I can see how that would tend to discretize distances. 

Thanks for being patient as I try to grok this!

> Are you thinking an interactive rebase and picking one of the rpy2 related commits and the final one?

Exactly!

grovduck commented:
@aazuspan, I've now updated the test files for those few records with tied distances.  This commit should be the one we keep to reflect the tie back to using R's `randomForest` objects.  From here out, we'll switch to `sklearn` objects.  If you get a chance to test that all tests pass in `test_port.py`, that would be great!

aazuspan commented:
Tests are passing on my end!

grovduck commented:
@aazuspan, got a bit waylaid with other tasks for the past couple of days, but also running up with issues proving analytically that we're getting similar results using R's `randomForest` regression and sklearn's `RandomForestRegressor`.  I wanted to pick your brain a bit and see if you think it's worth pursuing a test to prove correspondence or if we should just move on.

First off, the corresponding implementation using sklearn's `RandomForestRegressor` objects is fairly trivial.  For now, I have it separated into its own module, but will probably just put it in `RFNodeTransformer` once we're transitioned away from `rpy2` objects.  See `_rf_sklearn.py` for details.

I tried to make the configuration of the hyperparameters to be as close as possible between R and `sklearn`.  Right now, I am making these configurable and identical:
- R `n_tree` -> scikit `n_estimators`
- R `mtry` -> scikit `max_features`
- R's `nodesize` and scikit `min_samples_leaf` both default to 5
- Seed set to 42 for both

Based on a bit of searching (e.g. [here](https://stackoverflow.com/questions/35578041/why-do-i-get-different-randomforest-outcomes-between-r-and-python) and [here](https://stackoverflow.com/questions/73728280/too-good-to-be-true-results-using-sklearns-randomforestregressor-compared-with)), even with these configuration option set equal, the individual trees in the forests will differ from one another.  I don't think we are necessarily going to get around that.

The other issue is that our features in our test data (species basal area and density) have really poor fits to begin with, so my hypothesis is that these forests may vary pretty widely between the two implementations.

For now, I've coded `RFNodeTransformer` and `RFNNRegressor` with a `method` keyword that should either be `"rpy2"` or `"sklearn"` or order to test both at the same time.  I've also written two new tests:

1. Our original idea of testing correlations between `rpy2`- and `sklearn`-based estimators for each predicted feature.  While some features are not awful, most are much lower than my originally anticipated ~0.9 or so.  This test is called `test_high_correlation_rfnn` and I've set the expected correlations artificially high so that the print statement will show you the correlations between features
2. Knowing that predicted attributes tests won't pass until our neighbors match, the second test (`test_kneighbors_rfnn`) is currently running `k=13` models for both estimators and comparing their neighbors (k = 13 is completely arbitrary and is the most number of neighbors I could print out before it started wrapping output).  What I hope you'll see is that for each plot, the neighbor lists are close, but they often swap positions between two or three.  Here's an example of a high correspondence of nearest neighbors IDs between the two methods:
    ```
    [1002 1007 1403 1405 2004 1005 1503 1201 1508 1202 1206   35 1208]
    [1002 1007 1403 1405 2004 1503 1508 1201 1206 1005 1202 1204 1804]
    ```
    and here's an example of a relatively poor correspondence:
    ```
    [1801 1206 1503 2004 1204 1504 1211 1202 1002 2001 1804 1005 1806]
    [1804 1801 1206 1401 1203 1503 2004 1803 1205 1402 1202 1204 1002]
    ```
    This test will also fail at this point and print out the neighbor IDs for each method along with a couple of correspondence statistics (see below).

So even with the poor correspondence, there are a lot of similar neighbors in the top 13 neighbors, but they are further away from their position.  I'm using both Spearman's rank correlation and something I found called [Spearman's "footrule distance"](https://www.statisticshowto.com/spearman-footrule-distance/) to measure this correspondence, but I have no idea what a reasonable test value should be.

I think I'm fairly confident that the code is correct, but I'm struggling with how much more time we should spend on trying to get these two methods to agree before throwing up our hands.  If you have thoughts on any aspect of this approach, I'd love to hear it!  And please let me know if you need further explanation on what I've done so far - it's clear in my head, but I'm not sure I've written it very clearly.

aazuspan commented:
Thanks for the update and the deep dive into validation, @grovduck!  

> The other issue is that our features in our test data (species basal area and density) have really poor fits to begin with, so my hypothesis is that these forests may vary pretty widely between the two implementations.

That makes sense to me. I imagine correlation would be poor even if you compared the same implementation, i.e. multiple `sklearn` RF models with different random seeds - if the models can't converge towards an "accurate" final state, then it seems logical that minor differences could lead to big disagreements.

I'm happy to discuss further or try to come up with alternate validation approaches, but I'm more than satisfied with the validation you've already done. By testing this iteratively at each step, it seems like you've isolated the potential sources of error to 1) the tiny bit of code in fe095ec for instantiating and fitting the `sklearn` model, which is clearly correct, and 2) the expected differences between the `randomForest` and `sklearn` implementations of RF.

I'll follow your lead, but I have no reservations about moving ahead with this!

grovduck commented:
@aazuspan, I took your good advice and decided to move on without trying to match the output between R and scikit-learn.  This latest commit strips out all `rpy2` code, replaces the test output, and cleans up the code a bit.  I'll next take a first shot at making sure all tests pass along with setting tags and then I'd love it if you want to start finding the warts!

grovduck commented:
@aazuspan, I think I'm ready for your eyes on this.  It's currently passing all tests and I've added the first level of documentation, so it's probably close to a "minimum viable product".  However, there are still a number of features to address here and I'd love your perspective on these.

- This last commit, I've added `random_state` as a parameter to `RFNNRegressor` (and through to `RFNodeTransformer`).  I'm setting the seed to 42 in port tests to verify that this is still lining up.  However, there are *a lot* of kwargs on `RandomForestRegressor` that I don't know if we should surface all the way up to `RFNNRegressor`.  We have surfaced most of the parameters from `KNeighborRegressor`, so I guess we should do the same here?  Are there more elegant ways to do this other than parameter by parameter?

- The other obvious thing I need to do to make this estimator more useful is to have the set of random forests to be both `RandomForestClassifier` and `RandomForestRegressor` objects.  The `RandomForestClassifier` forests could either be built from classifying continuous `y` attributes into classes or from `y` features that already represent classes, either numeric or strings.  In my mind, this seems like we need to both: 1) tell the estimator how to build each forest (some kind of mapper between `y` attribute and random forest type); and 2) have utility functions to bin continuous data into categorical data.  I can definitely see the argument that 2. should be outside the scope of the estimator such that the user does this on their own, but it might be nice to have example utility functions to show how this is done.  

I don't know if it makes sense to finish up this PR given that we've replicated `yaImpute` and handle the above (and other) features in separate issues/PRs or just continue working in this space.  Other than that, any other comments that you see on the existing code would be welcome.

aazuspan commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -0,0 +1,105 @@
+        if y.ndim == 1:
+            y = y.reshape(-1, 1)
+
+        self.rfs_ = [
```
We can certainly tackle this separately, but I wonder if we should parallelize this with `joblib` based on `n_jobs`? I did a quick test fitting with SWO Ecoplot, the default 50 trees, and `n_jobs=-1` and the time went from 14.1s as-is to 8.7s when parallelizing this loop and passing `n_jobs` to the RF constructor. Creating the forests serially but passing `n_jobs` to the RF constructor took 9.6s.

grovduck replied:
This sounds like a great idea.  I think it would look something like this, but I'm not nearly as familiar as you are with it:
```
from joblib import Parallel, delayed

self.rfs_ = Parallel(n_jobs=self.n_jobs)(
    delayed(lambda i: RandomForestRegressor(
        n_estimators=self.n_estimators_per_forest,
        max_features=self.max_features,
        random_state=self.random_state,
        min_samples_leaf=5,
        n_jobs=self.n_jobs,
    ).fit(X, y[:, i]))(i)
    for i in range(y.shape[1])
)
```
It did seem to slow down tests.  Does that seem right?

grovduck replied:
Yeah, it's strange.  Obviously in tests, we are instantiating these objects quite a bit and there is presumably overhead associated with setting up `joblib`, but I can't reconcile why this counters the speedups that you saw.  I'm OK pushing this to another PR, although you might be the one doing the heavy lifting!

aazuspan replied:
You may be right about the joblib overhead causing the slowdown - hopefully this isn't another mystery threading issue, although I could see some strangeness arising from parallelizing the already parallel fitting of RFs... Worth a closer look, for sure.

aazuspan reviewed (changes requested):
This looks great - I'm really excited to have a working RF-NN in Python! And with all the effort you put into validating along the way, I think we can be very confident in "correctness" relative to `yaImpute`.

I have a few minor nits, as well as some other comments that we can tackle after merging, if you want. I do think we should decide on how best to handle the RF parameters in this PR, whether we list them all explicitly or unpack them from an optional dictionary.

We should also update the readme to 1) remove the footnote that notes RF-NN is in development and 2) possibly explain the random forest implementation discrepancy in the "validation" footnote.

> The other obvious thing I need to do to make this estimator more useful is to have the set of random forests to be both RandomForestClassifier and RandomForestRegressor objects.

Given that we have a working version of RF-NN without this feature, I would suggest we split this out into a separate PR. It seems like it could be pretty involved, depending on how we choose to tackle it.

grovduck commented on tests/test_estimators.py:
```diff
@@ -91,7 +94,7 @@ def get_estimator_xfail_checks(estimator) -> dict[str, str]:
         )
 
-    if isinstance(estimator, (GNNRegressor, MSNRegressor)):
+    if isinstance(estimator, (GNNRegressor, MSNRegressor, RFNNRegressor)):
```
So this would include both `EuclideanKNNRegressor` and `MahalanobisKNNRegressor` which wouldn't fail because input features equal transformed features?  But I guess we want to be clear that a `TransformedKNeighborsRegressor` subclass *may* have a different number of input and transformed features.  I'll make the change to `TransformedKNeighborsRegressor`, but wanted to be clear that we're xfailing tests that would pass.

aazuspan replied:
I totally overlooked that this includes more estimators than necessary. I think you're right that xfailing passing tests is the wrong approach here. 

I was thinking that any transformation would mean that input features are mismatched, but of course that's only the case when the transformer changes the number/names of features. I think you can ignore my comment and leave this as-is.

grovduck replied:
OK, sounds good.  I was thinking that I would use `Y_FIT_ESTIMATORS` instead, but that suggests that there is something special about having a `y_fit` parameter that makes the number of input features not equal the number of transformed features.   Clearly just using a `PCA` without any `y` would have the same behavior.  I'll keep it as-is.

grovduck commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -0,0 +1,105 @@
+        self.max_features = max_features
+        self.random_state = random_state
+
+    def fit(self, X, y=None):
```
Yep, good catch!

grovduck commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -0,0 +1,105 @@
+        check_is_fitted(self, "rfs_")
+        return np.asarray(
+            [
+                f"node_{i}_{j}"
```
I like your names of the features much better than mine.  I've updated to use your names.

grovduck commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -0,0 +1,105 @@
+    def __sklearn_tags__(self):
+        tags = super().__sklearn_tags__()
+        tags.target_tags.required = True
+        tags.transformer_tags.preserves_dtype = []
```
I think you're right - I was confused by the description [here](https://scikit-learn.org/stable/modules/generated/sklearn.utils.TransformerTags.html#sklearn.utils.TransformerTags).  I was thinking I needed to provide an empty list because input features could be dtypes other than `int64` and thus would not preserve dtypes.  But the last sentence clarifies that the first value listed is the default when it's not expected to preserve dtypes.  Thanks for catching this.

grovduck commented on src/sknnr/_rfnn.py:
```diff
@@ -0,0 +1,111 @@
+        weights: Literal["uniform", "distance"] | Callable = "uniform",
+        algorithm: Literal["auto", "ball_tree", "kd_tree", "brute"] = "auto",
+        leaf_size: int = 30,
+        n_jobs: int | None = None,
```
> I do think we should expose every RF param _somehow_ to allow full control, whether they're explicit or passed through a param dict. Listing them all explicitly would provide the best documentation and code completion, but it would also require a ton of boilerplate and need to be maintained to match `sklearn`. Including the important ones explicitly (like you've already done) and allowing extras via an `rf_params: dict` or similar seems like a reasonable compromise to me, but I'm happy whichever way you prefer to go.

I think I'm leaning toward your `rf_params` idea, but wanted to sketch it out before committing.  Because we're relying on both `RandomForestRegressor` and `KNeighorsRegressor` here, I think it makes sense to lump the kwargs associated with each in the parameter signature.  It feels strange to me to list the kneighbor params explicitly, but rely on a dict to hold the rf params.  We could do something like:
```
def __init__(
    self,
    n_estimators_per_forest: int = 50,
    n_neighbors: int = 5,
    *,
    rf_params: dict[str, Any] = None,
    kneighbors_params: dict[str, Any] = None
)
```
This would surface the parameters that are not required to be keywords (`n_estimators_per_forest` and `n_neighbors`) and put everything else into the two dicts. But it might leave out defaults for parameters that we consider "important".

Because we're using Hamming distance, the kneighbors parameters `p` and `metric` would have no effect so including them in `kneighbors_params` would also have no effect.  Is that an issue?  I don't think so, because presumably including `p` with `metric="hamming"` in `KNeighborsRegressor` would also have no effect.

> Either way, we'll need to be careful with how we handle 1) renamed parameters like `n_estimators_per_forest` and 2) parameters that can be used by both random forests and kNN, like `n_jobs`.

For 1), I went back and forth on calling this parameter `n_estimators`.  That would certainly be cleaner, but I wondered if users would get confused whether this represented the total number of estimators (across all forests) or estimators per forest.  I'm totally OK renaming this if you think `n_estimators` is more clear.

For 2), if we went with dicts to pass the additional params to each sub-estimator, this wouldn't be an issue.  I guess I'm a little leery to have parameters like `n_jobs_rf` and `n_jobs_knearest`.

Thoughts?

aazuspan replied:
>  Because we're relying on both RandomForestRegressor and KNeighorsRegressor here, I think it makes sense to lump the kwargs associated with each in the parameter signature. It feels strange to me to list the kneighbor params explicitly, but rely on a dict to hold the rf params. 

Agreed, I like your design separating `rf_params` and `kneighbors_params`.

> This would surface the parameters that are not required to be keywords (n_estimators_per_forest and n_neighbors)

I'm curious about including these two explicitly instead of as RF and KNN params, respectively. I'm not opposed to that, but just wondering if that was done for convenience since these are the most likely parameters to adjust, or to make the defaults clear? We *could* store default versions of both param dicts and update them with the user dicts to handle default values (being careful not to override the critical ones like `metric`), but that does have drawbacks like defaults being less clear and making it harder to adjust common parameters.

> Because we're using Hamming distance, the kneighbors parameters p and metric would have no effect so including them in kneighbors_params would also have no effect. Is that an issue? I don't think so, because presumably including p with metric="hamming" in KNeighborsRegressor would also have no effect.

I think I agree with your conclusion here. We could note in the docstring that some parameters are ignored, but as you mention that's not unusual for estimators.

> For 1), I went back and forth on calling this parameter n_estimators. That would certainly be cleaner, but I wondered if users would get confused whether this represented the total number of estimators (across all forests) or estimators per forest. I'm totally OK renaming this if you think n_estimators is more clear.

I agree with that `n_estimators_per_forest` is clearer as a parameter to `RFNNRegressor`, but I'm a little unsure what would/should happen if a user passed `rf_params=dict(n_estimators=100))`? 

> For 2), if we went with dicts to pass the additional params to each sub-estimator, this wouldn't be an issue. I guess I'm a little leery to have parameters like n_jobs_rf and n_jobs_knearest.

Good point that passing separate dicts allows full control over these, and I agree that separating them by name would be messy. Another option would be to explicitly accept the parameters that are shared by both estimators like `n_jobs` (maybe that's the only one?) and then pass them to both estimators (override-able by the respective dicts), but I don't have strong feelings on that. It's easy enough for users to just specify both.

grovduck replied:
@aazuspan, this one turned into a hornet's nest (at least the way I was trying to implement it).  I wonder if you might have suggestions that get around some of the issues that I've found:

- There is a caution in the [custom estimator guide](https://scikit-learn.org/stable/developers/develop.html) that warns against using mutable parameters (like dicts) and that, if they are used, they should be copied in `__init__` but nothing else.  That means stitching a user-passed dict (like `rf_params`) along with a default dict (defined in `__init__`) causes issues in some of the tests because there are cloning issues (unfortunately, I didn't capture the error message exactly, but I believe this happens when testing this estimator in pipelines.  I can probably recreate the test errors if needed.)
- Your concern with holding out `n_neighbors` and `n_estimators_per_forest` is completely valid - I agree that this is a strange pattern.  I did this mostly to avoid issues with our port tests as all of our other estimators can accept a `n_neighbors` parameter whereas we would be burying this `n_neighbors` key in the `kneighbors_params` dictionary if we completely separated them.  My preference is that any of our estimators take the `n_neighbors` parameter to enforce the idea that this is truly a set of k-neighbors estimators.
- For the `RFNodeTransformer`, if we pass an `rf_params` dictionary instead of all the allowed `RandomForestRegressor` keywords, we run into issues with idempotency (e.g. `est.fit(X, y)` has the same result as `est.fit(X, y).fit(X, y)`).  The way I was implementing it, I just set `self.rf_params = rf_params.copy()` in `__init__` and then in `fit`, I stitched together the user-passed params and a default params dictionary.  If `random_state` is `None` here, then I'm creating a new forest for each `y` attribute with a new random seed.  You can see in the [`check_fit_idempotent` function](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/estimator_checks.py#L4263-L4275) that the test sets the random state after initialization to 0, but this would have no effect, because `fit` would be after this.

So, you can see that I've completely punted on this latest commit and added all keywords from _both_ `RandomForestRegressor` and `KNeighborsRegressor`.  As you noted, this is a lot of boilerplate, but as far as I can tell, it seems like this isn't unusual as a pattern in `scikit-learn`.  Admittedly, I'm a bit "done with this", but if you see opportunities that I'm missing, I'd love your feedback or suggested code.  

Even with this commit, there are a few minor issues:
- Right now, I still only have a single `n_jobs` keyword argument and this is being used for both the RF and KNN.  We can revisit providing two parameters for this and customizing their names (BTW, `n_jobs` _is_ the only parameter that both estimators share.)
- My typing on a couple of parameters likely isn't correct, specifically: 1) `ccp_alpha` which is supposed to be a _non-negative_ float and 2) `monotonic_cst` which is "array-like of int of shape (n_features)".  I don't think there is specific type information in the code for this parameter, only in the docstring.

aazuspan replied:
I forgot that `sklearn` is really strict about putting logic into `__init__` (presumably that would break `get_params` and `set_params`). Duplicating and typing all the params from sklearn must have been a pain, but I agree with your choice to go that route rather than trying to come up with a workaround for the param dicts. Plus, it'll be great for code completion and documentation.

> Right now, I still only have a single n_jobs keyword argument and this is being used for both the RF and KNN.

I can't think of a good reason why someone would want to run one in parallel but not the other, so I think a shared parameter is probably fine.

> My typing on a couple of parameters likely isn't correct, specifically: 1) ccp_alpha which is supposed to be a non-negative float and 2) monotonic_cst which is "array-like of int of shape (n_features)". I don't think there is specific type information in the code for this parameter, only in the docstring.

The only way I know to handle these kind of complex type constraints is with runtime checks and `typing.TypeGuard`, but I'm not sure that's worth the effort. It might also open a can of worms about which other parameter constraints we need to validate...

I need to give the code another careful look, but as far as I'm concerned it's in a great state and seems like it's just about wrapped up!

grovduck replied:
Thanks for the feedback, @aazuspan.  Once last thing that I failed to mention is that I went back to `n_estimators` rather than `n_estimators_per_forest` to keep it totally consistent with `RandomForestRegressor`.  I think we can use the docstring to make clear that this applies per forest.

If we've settled on explicit listing of all parameters for now, I'll fix up docstrings and make another commit.

aazuspan commented on tests/test_port.py:
```diff
@@ -44,6 +46,9 @@ def test_kneighbors(result, n_components):
     hyperparams.update(
         {"n_components": n_components} if hasattr(estimator(), "n_components") else {}
     )
+    hyperparams.update(
+        {"random_state": 42} if hasattr(estimator(), "random_state") else {}
+    )
```
Sounds good to me! 

I played around with refactoring this when I did the first review and used a helper function that allows overriding the common default params and removes params that doesn't fit the estimator's signature. Feel free to pull from this if it's helpful or ignore it if you had a different plan! 

```python
import inspect

def get_default_hyperparams(estimator, **kwargs) -> dict:
    """Return valid parameters for the given estimator, including common defaults."""
    default_params = dict(
        n_neighbors=5,
        random_state=42,
        **kwargs,
    )

    valid_params = inspect.signature(estimator).parameters
    return {k: v for k, v in default_params.items() if k in valid_params}
```

grovduck replied:
As always, I will gladly take your work whenever I can!  This worked perfectly, thx.

grovduck commented:
@aazuspan, I *think* I've addressed all the outstanding issues on this one:

- Added your helper function to help simplify hyperparameters
- Copied docstring information from `RandomForestRegressor` and `KNeighborsRegressor` into `RFNNRegressor` and `RFNodeTransformer`.  For each docstring, I just copied the initial sentence from the `sklearn` documentation and referred the user to the source `sklearn` classes for more detail
- Added new stubs in the documentation for these classes
- Expanded the usage doc a bit to explain distance in RFNN space

No worries if you can't review today.  This can sit.

aazuspan reviewed (changes requested):
@grovduck, just a couple final typo/formatting nits in the docs, but otherwise LGTM! The extra details you added to the docs look great, and I like your decision to give basic docstrings and point to sklearn for more info. This was a big lift, but it came out great and I'm excited to try mapping with it!

I'd say go ahead and merge whenever you're happy!

grovduck commented on README.md:
```diff
@@ -85,5 +85,4 @@ Thanks to Andrew Hudak (USDA Forest Service Rocky Mountain Research Station) for
-[^rfnn]: In [development](https://github.com/lemma-osu/sknnr/issues/24)!
-[^validation]: All estimators and parameters with equivalent functionality in `yaImpute` are tested to 3 decimal places against the R package.
\ No newline at end of file
+[^validation]: All estimators and parameters with equivalent functionality in `yaImpute` are tested to 3 decimal places against the R package. Because of implementation differences between R's `randomForest` package and scikit-learn's `RandomForestRegressor`, the RFNN estimator is not directly comparable to the `yaImpute` implementation. However, in porting RFNN to `sknnr`, we ensured that we obt ...
```
Good catch!

grovduck commented on docs/pages/usage.md:
```diff
@@ -62,16 +63,18 @@ print(y.loc[neighbor_ids[0]])
 
 !!! tip
-    In forestry applications, users typically store a unique inventory plot identification number as the index in the dataframe.
+In forestry applications, users typically store a unique inventory plot identification number as the index in the dataframe.
```
That was auto-formatting and me committing those changes not realizing the impact!  Thanks for catching that.  All reverted now.

aazuspan reviewed (approved):
LGTM!

grovduck commented:
@aazuspan Awesome, thanks for all your help getting this implemented!  It's very helpful having your insight on design issues.

grovduck commented:
> > > As a side note, rather than squashing all commits on this PR, I wonder if it might be worth holding onto one with the rpy2 implementation so that we can come back to it at some point?
> > 
> > 
> > Yeah, rebasing this one makes sense to me.
> 
> I might need some help with this one when it comes time to do so. Are you thinking an interactive rebase and picking one of the `rpy2` related commits and the final one? We can cross that bridge when we get there.

@aazuspan - D'oh!  I totally forgot to do this when I committed this.  I'm not sure it's really worth going back into this as we have the commit history in this PR.  I'm sure I would bungle something if I tried to redo at this point.

aazuspan commented:
Yeah, I'm not sure how to do that without potentially making a mess downstream... Oh well. As you said, we've got the history here at least.
